### PR TITLE
Support DP MLA

### PR DESCRIPTION
**Pull request by ispobock**
## Motivation

Support data parallel on MLA for DeepSeek model to reduce replicated KV cache. 
```
python -m sglang.launch_server --model-path neuralmagic/DeepSeek-Coder-V2-Instruct-FP8 --trust-remote-code --tp 8 --dp 8 --enable-dp-attention
```
## Modifications
- Add `--enable-dp-attention` option. When it is turned on, DP and TP share the same workers.
- Add `IDLE` forward mode for workers that do not have sequence to forward but need TP sync with other workers.
- Implement model forward with DP attention + TP MoE.

## Performance
Compared to the main branch, this PR improves the prefill throughput by **20%** and the decode throughput by **67%** for DeepSeek-V2 model on 8*H100.

DP+TP (this PR):
- prefill: 21658.78 toks/s
- decode: 11174.62  toks/s

TP (main branch):
- prefill: 17941.92 toks/s
- decode: 6656.75 toks/s

Reproduce:
```bash
# DP+TP
python -m sglang.launch_server --model-path neuralmagic/DeepSeek-Coder-V2-Instruct-FP8 --disable-radix-cache --trust-remote-code --tp 8 --dp 8 --enable-dp-attention
# TP
python -m sglang.launch_server --model-path neuralmagic/DeepSeek-Coder-V2-Instruct-FP8 --disable-radix-cache --trust-remote-code --tp 8

# bench prefill
python3 -m sglang.bench_serving --backend sglang --dataset-name random --random-input 512 --random-output 1 --random-range-ratio 1 --num-prompts 10000
# bench decode
python3 -m sglang.bench_serving --backend sglang --dataset-name random --random-input 1 --random-output 512 --random-range-ratio 1 --num-prompts 10000
```

## TODO
- Compatible with cuda graph: https://github.com/sgl-project/sglang/pull/2061
- Compatible with overlap mode

### CODE REVIEW

**Comment by fengyang95** [comment]
@ispobock  How much performance improvement is expected? Is it mainly in throughput or latency?

**Comment by ispobock** [comment]
> How much performance improvement is expected? Is it mainly in throughput or latency?

@fengyang95 There is an issue with dp 8. I will test the performance once the issue fixed. It's mainly for throughput.

**Comment by fengyang95** [comment]
> * Compatible with cuda graph
> * Compatible with overlap mode

@ispobock hi, when will support for cuda graph be planned? It is critical for latency improvement.

File: python/sglang/srt/server_args.py [hunk 1]
```diff
@@ -193,6 +194,10 @@ def __post_init__(self):
             self.disable_penalizer = True
             self.disable_nan_detection = True
 
+        if self.enable_dp_mla:
+            self.dp_size = self.tp_size
+            self.chunked_prefill_size = 4096
```
Comment (merrymercy) [inline python/sglang/srt/server_args.py]: print a warning here or use `self.chunked_prefill_size /= 2` so the user can still control this.

File: python/sglang/srt/models/deepseek_v2.py [hunk 2]
```diff
@@ -491,6 +532,41 @@ def forward(
         return output
 
 
+def all_gather(input_tensor: torch.Tensor, rank, world_size, group):
+    if world_size == 1:
+        return input_tensor
+
+    seq_len = input_tensor.size(0)
+    local_len = torch.tensor([seq_len], dtype=torch.int64, device=input_tensor.device)
```
Comment (merrymercy) [inline python/sglang/srt/models/deepseek_v2.py]: I think this also works
```suggestion
    local_len = torch.tensor(seq_len, dtype=torch.int64, device=input_tensor.device)
```

File: python/sglang/srt/models/deepseek_v2.py [hunk 3]
```diff
@@ -646,10 +745,18 @@ def __init__(
         self.config = config
         self.quant_config = quant_config
         self.model = DeepseekV2Model(config, cache_config, quant_config)
-        self.lm_head = ParallelLMHead(
-            config.vocab_size, config.hidden_size, quant_config=quant_config
-        )
-        self.logits_processor = LogitsProcessor(config)
+        if global_server_args_dict["enable_dp_mla"]:
+            self.lm_head = ReplicatedLinear(
```
Comment (merrymercy) [inline python/sglang/srt/models/deepseek_v2.py:744]: why cannot we parallelize this?

File: python/sglang/srt/managers/scheduler.py [hunk 4]
```diff
@@ -346,12 +354,17 @@ def event_loop_normal(self):
...
                             break
+                        if self.server_args.enable_dp_mla:
+                            self.check_dp_batch_running(batch)
                         result = self.run_batch(batch)
                         self.process_batch_result(batch, result)
             else:
                 self.check_memory()
                 self.new_token_ratio = self.init_new_token_ratio
 
+            if batch and batch.forward_mode.is_idle():
```
Comment (merrymercy) [inline python/sglang/srt/managers/scheduler.py]: why is this needed?

File: python/sglang/srt/managers/schedule_batch.py [hunk 5]
```diff
@@ -56,6 +56,7 @@
     "disable_mla": ServerArgs.disable_mla,
     "torchao_config": ServerArgs.torchao_config,
     "disable_nan_detection": ServerArgs.disable_nan_detection,
+    "enable_dp_mla": ServerArgs.enable_dp_mla,
```
Comment (merrymercy) [inline python/sglang/srt/managers/schedule_batch.py]: I feel this is quite general, can we call it `enable_dp_attention`?

File: python/sglang/srt/managers/scheduler.py [hunk 6]
```diff
@@ -332,6 +332,14 @@ def event_loop_normal(self):
             self.process_input_requests(recv_reqs)
 
             batch = self.get_next_batch_to_run()
+
+            if (
+                self.server_args.enable_dp_mla
+                and self.check_dp_batch_running(batch)
+                and batch is None
+            ):
+                batch = self.get_idle_batch()
```
Comment (merrymercy) [inline python/sglang/srt/managers/scheduler.py]: can we move this logic inside `self.get_next_batch_to_run()`?

File: python/sglang/srt/managers/scheduler.py [hunk 7]
```diff
@@ -346,12 +354,17 @@ def event_loop_normal(self):
                         self.update_running_batch()
                         if not self.running_batch:
                             break
+                        if self.server_args.enable_dp_mla:
+                            self.check_dp_batch_running(batch)
```
Comment (merrymercy) [inline python/sglang/srt/managers/scheduler.py]: It seems the return value is not used.

File: python/sglang/srt/server_args.py (hunk 1 above)
Comment (ispobock) [reply python/sglang/srt/server_args.py]: Make sense.

File: python/sglang/srt/models/deepseek_v2.py (hunk 2 above)
Comment (ispobock) [reply python/sglang/srt/models/deepseek_v2.py]: Let me verify this.
I will move this part to scheduler, because we know the num_tokens since we start the batch and it's static for each iteration. We don't need to sync it for each layer.

File: python/sglang/srt/models/deepseek_v2.py (hunk 3 above)
Comment (ispobock) [reply python/sglang/srt/models/deepseek_v2.py:744]: Because the data in each worker is different. We need to all-gather data before the TP and slice data after it. Extra operations and communication overhead are added.

File: python/sglang/srt/managers/scheduler.py (hunk 7 above)
Comment (ispobock) [reply python/sglang/srt/managers/scheduler.py]: Yes, we just need a sync here before the run_batch.

File: python/sglang/srt/managers/scheduler.py (hunk 6 above)
Comment (ispobock) [reply python/sglang/srt/managers/scheduler.py]: Make sense.

File: python/sglang/srt/managers/schedule_batch.py (hunk 5 above)
Comment (ispobock) [reply python/sglang/srt/managers/schedule_batch.py]: Sure.

File: python/sglang/srt/managers/scheduler.py (hunk 4 above)
Comment (ispobock) [reply python/sglang/srt/managers/scheduler.py]: Avoid to change the last batch here. But the last batch seems not used in the normal loop. We can remove this check.

**Comment by ispobock** [comment]
> when will support for cuda graph be planned?

I will support it soon. The code is almost done and needs some tests.

**Comment by Xu-Chen** [comment]
@ispobock 
Is there a plan to support A100 or GPTQ quantized models?

**Comment by fengyang95** [comment]
> > when will support for cuda graph be planned?
> 
> I will support it soon. The code is almost done and needs some tests.

@ispobock How much additional VRAM would this require approximately?

File: python/sglang/srt/managers/scheduler.py (hunk 6 above)
Comment (ispobock) [reply python/sglang/srt/managers/scheduler.py]: The code is updated. There are many return branchs in `get_next_batch_to_run()` and the idle case should be handled based on the returned batch. So I think it's easier to handle it out of `get_next_batch_to_run()`.

**Comment by ispobock** [comment]
> How much additional VRAM would this require approximately?

@fengyang95 For the 236B V2 model, if use DP attention, the total weights will take ~570G, so prefer to use FP8 quantized model for better performance.

**Review by merrymercy — APPROVED** [review]
Great work!

**Comment by fengyang95** [comment]
@ispobock Does this support W4A16? My VRAM is very limited, and even using fp8 VRAM is not enough.

**Comment by ispobock** [comment]
>  Does this support W4A16?

@fengyang95 AWQ is supported in https://github.com/sgl-project/sglang/pull/2364.

**Comment by faradawn** [comment]
Hi, for `--tp 8 --dp 8 --enable-dp-attention`, 1) is it that we replicate the attention layer (the weights W^Q, W^KV) 8 times? 2) Is it that we pass different data to each TP worker? 3) If without dp attention, would TP=8 and DP=8 require 64 GPUs?

**Comment by chenhao-stick-to** [comment]
> 您好，关于 TP 工作线程`--tp 8 --dp 8 --enable-dp-attention`，1）是否需要将注意力层（权重 W^Q、W^KV）复制 8 次？2）是否需要将不同的数据传递给每个 TP 工作线程？3）如果不使用动态规划注意力机制，TP=8 和 DP=8 是否需要 64 个 GPU？

关于这种情况，结合dp attention的真实目的是减少mla模型重复kvcache缓存以及底层其实是复用一个tp group的实现来看，我说说我的看法（如有不对处，欢迎一起讨论）：
1）不需要，实际上还是会将完整注意力层切成8份（按照tp8操作），然后计算后再all gather到对应的逻辑dp rank，这里显然有8个逻辑dp rank，所以其实只是gather到对应的dprank
2）据我所了解的情况来看，是的，需要这么多GPU。
